lib/index.js

```diff
@@ -11,89 +11,95 @@ const Manager = require('./manager');
 
 // Declare internals
 
-const internals = {};
+const internals = { schema: {} };
 
+internals.schema.manager = Manager.schema;
 
-internals.schema = Joi.alternatives([
+internals.schema.handler = [
     Joi.string(),
     Joi.object({
         template: Joi.string(),
         context: Joi.object(),
         options: Joi.object()
     })
-]);
-
+];
 
 exports.plugin = {
-    once: true,
-    pkg: require('../package.json'),
-
-    register: function (server, pluginOptions) {
 
-        server.decorate('server', 'views', function (options) {
+    multiple: true,
+    pkg: require('../package.json'),
 
-            Hoek.assert(options, 'Missing views options');
-            this.realm.plugins.vision = this.realm.plugins.vision || {};
-            Hoek.assert(!this.realm.plugins.vision.manager, 'Cannot set views manager more than once');
+    register: function (server, options) {
 
-            if (!options.relativeTo &&
-                this.realm.settings.files.relativeTo) {
+        if (Object.keys(options).length > 0) {
+            const boundAssignManager = internals.assignManager.bind(server);
+            boundAssignManager(options);
+        }
 
-                options = Hoek.shallow(options);
-                options.relativeTo = this.realm.settings.files.relativeTo;
-            }
+        const rootRealm = internals.getRootRealm(server.realm);
+        const rootState = internals.state(rootRealm);
 
-            const manager = new Manager(options);
-            this.realm.plugins.vision.manager = manager;
-            return manager;
-        });
+        if (rootState.setup) {
+            return;
+        }
 
+        server.decorate('server', 'views', internals.assignManager);
         server.decorate('server', 'render', internals.render);
         server.decorate('request', 'render', internals.render);
         server.decorate('handler', 'view', internals.handler);
+        server.decorate('toolkit', 'view', internals.toolkitView);
+
+        rootState.setup = true;
+    }
+};
 
-        server.decorate('toolkit', 'view', function (template, context, options) {
 
-            const realm = (this.realm.plugins.vision || this.request.server.realm.plugins.vision || {});
-            Hoek.assert(realm.manager, 'Cannot render view without a views manager configured');
-            return this.response(realm.manager._response(template, context, options, this.request));
-        });
+internals.assignManager = function (options) {
+
+    Joi.assert(options, internals.schema.manager, 'Invalid manager options');
+
+    const realm = this.realm.parent || this.realm;
```

Comment:
Wait, when `server.views()` is called inside a plugin wont this set the manager on the plugin's parent?  It would be hard to see this in a test when `nearestState()` comes into play.  But imagine `server1` registers a plugin with `server2`, and `server2` register a plugin with `server3`.  Now `server3` calls `server3.views()`– I think this will cause `server2` and `server3` to both have the same manager.

Reply:
I think you're probably right, I'll definitely go back and ensure this acts correctly in tests